Add global cache to map between the deviceclass and the extended resource

sairameshv opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
/wg device-management

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR adds a global cache for the event handler functions of the noderesources plugin to look into the mapping of the device class and the associated DRA extended resource

#### Which issue(s) this PR is related to:
Fixes https://github.com/kubernetes/kubernetes/issues/133670
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [DRA Extended resources]: <https://github.com/kubernetes/enhancements/issues/5004>
- [KEP]: <https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/5004-dra-extended-resource#readme>
- [Issue]: <https://github.com/kubernetes/kubernetes/issues/133670>
```

cc: @yliaog

bart0sh commented on pkg/scheduler/framework/plugins/dynamicresources/extended/cache.go:
```diff
@@ -0,0 +1,126 @@
+type ExtendedResourceCache struct {
+	mutex sync.RWMutex
+	// mapping maps extended resource name to device class name
+	mapping    map[v1.ResourceName]string
```
JFYI: There is [ongoing debate](https://github.com/kubernetes/kubernetes/pull/133734/files#r2347169756) about changing this mapping to `map[v1.ResourceName]*resourceapi.DeviceClass`.

sairameshv commented:
> @sairameshv thanks for the PR. please remove Draft status from PR when it is ready for review.

Opened the PR for the review. Thanks

sairameshv commented:
/test pull-kubernetes-e2e-gce

yliaog commented on pkg/scheduler/framework/plugins/dynamicresources/extended/cache.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+// GetAllMappings returns a copy of all current mappings.
+func (c *ExtendedResourceCache) GetAllMappings() map[v1.ResourceName]string {
```
i think we can just use GetDeviceClass() to read directly from the cache, w/o needing to store a copy of it in noderesources and dynamicresource plugin.

sairameshv replied:
I remember adding this method for some reason while resolving some of the test case failures. But I got that working with this method now.

yliaog commented on pkg/scheduler/framework/plugins/dynamicresources/extended/cache.go:
```diff
@@ -0,0 +1,126 @@
+
+	case watch.Deleted:
+		// Remove from mapping
+		for resourceName, className := range c.mapping {
```
what about something similar to the Added, Modified, i.e., use deviceClass.Spec.ExtendedResourceName and resourceapi.ResourceDeviceClassPrefix + deviceClass.Name as key to delete from the map, instead of looping over the map?

sairameshv replied:
Addressed

alaypatel07 commented on pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:
```diff
@@ -263,3 +294,42 @@ func (c *claimTracker) AssumeClaimAfterAPICall(claim *resourceapi.ResourceClaim)
 }
+
+// deviceClassEventHandler handles device class informer events to update the extended resource cache.
+type deviceClassEventHandler struct {
```
Instead of introducing a new type just for registering the event handlers, can we make `ExtendedResourceCache` implement these event handlers: OnAdd, OnUpdate and OnDelete? I think the abstractions will be simpler that way

sairameshv commented on staging/src/k8s.io/kube-scheduler/framework/listers.go:
```diff
@@ -108,6 +109,17 @@ type ResourceClaimTracker interface {
+	// GetAllMappings returns a copy of all current mappings.
+	GetAllMappings() map[v1.ResourceName]string
+	// Refresh manually triggers a cache refresh.
+	Refresh()
```
This was added keeping in my mind the fact that there could already be a number of device classes that exist before initializing the cache and this would provide us a way to populate the cache by listing all the existing device classes.
I agree this need not be part of the interface and I limited this method's scope to the `cache.go`

sairameshv commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1720,3 +1743,9 @@ func getAllocatedDeviceStatus(claim *resourceapi.ResourceClaim, deviceRequest *r
+// getDeviceClassMappingFallback provides a fallback mechanism to get device class mapping
+// when the cache is not populated. This is mainly for test compatibility.
+func (pl *DynamicResources) getDeviceClassMappingFallback() (map[v1.ResourceName]string, error) {
+	return extended.DeviceClassMapping(pl.draManager)
```
I feel the fallback is needed when the cache is initialized and no yet populated before the informer caches are synced. Hence, I modified to use the existing cache and only fallback incases where the cache is not yet populated

sairameshv commented:
Hello reviewers, I have addressed the review comments and am planning to start working on the aspects of performance [benchmark](https://github.com/kubernetes/kubernetes/issues/133670#issuecomment-3365414479) discussed on the related issue by @bart0sh  Meanwhile could you all PTAL at the code once again? The change is more simplified now from its previous version

Thanks

yliaog commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -190,6 +187,24 @@ func NewFit(_ context.Context, plArgs runtime.Object, h fwk.Handle, fts feature.
 type ResourceRequestsOptions struct {
 	EnablePodLevelResources   bool
 	EnableDRAExtendedResource bool
+	DRAManager                fwk.SharedDRAManager
```
DRAManager  is not a ResourceRequestsOptions, so it seems weird to put it here. I can see it's convenient to have it here, so it can be passed around without fewer changes to isFit -> Fits -> fitRequest signature. I'll leave it to @macsko to decide. But at least add some comments to explain why it is here.

sairameshv replied:
Addressed

yliaog commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache_test.go:
```diff
@@ -0,0 +1,99 @@
+	cache.OnDelete(deviceClass1)
+	deviceClassName = cache.GetDeviceClass("nvidia.com/gpu")
+	if deviceClassName != "" {
+		t.Errorf("Expected 'nvidia.com/gpu' to be removed after deleting device class, got %s", deviceClassName)
```
try to avoid specific vendor, use example.com instead

yliaog commented:
@macsko could you take a look at this PR? there are a few other PRs pending on this, so it's better to have a resolution on this PR, so we have some time for other PRs to rebase before the code freeze.

macsko commented on pkg/scheduler/framework/plugins/noderesources/fit_test.go:
```diff
@@ -693,9 +688,15 @@ func TestEnoughRequests(t *testing.T) {
 			defer cancel()
 
+			if test.draExtendedResourceEnabled {
+				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DRAExtendedResource, true)
```
Move it just below `t.Run` and `if` can be omitted:
```suggestion
				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DRAExtendedResource, test.draExtendedResourceEnabled)
```

yliaog commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -632,17 +628,8 @@ func fitsRequest(podRequest *preFilterState, nodeInfo fwk.NodeInfo, ignoredExten
-			_, okScalar := nodeInfo.GetAllocatable().GetScalarResources()[rName]
-			_, okDynamic := podRequest.resourceToDeviceClass[rName]
-
-			if (okDynamic || podRequest.resourceToDeviceClass == nil) && !okScalar {
```
i think after we have this global mapping, the mapping should never be nil. i asked a few times in the last review, when would the mapping be nil? 

i agree it should keep the prior behavior, i.e., satisfied, pass to DRA dynamic resources plugin if the mapping happen to be nil.

sairameshv replied:
> Since this wasn't caught by any test, you might try to write one to make sure it won't be missed easily in the future. Edit: Or maybe it was caught by the e2e test

Yes, this is getting caught by a testcase already and hence the [failure](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134326/pull-kubernetes-kind-dra-all/1981281252140912640)

macsko commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -648,19 +648,27 @@ func addAllEventHandlers(
 }
 
-func preCheckForNode(nodeInfo *framework.NodeInfo) queue.PreEnqueueCheck {
+func (sched *Scheduler) preCheckForNode(nodeInfo *framework.NodeInfo) queue.PreEnqueueCheck {
```
> I think you mentioned 1/ is not really used anymore, is that correct?

Right, 1/ is not used.

> Given these, i think AdmissionCheck() does not need to use the global cache of extended resource to device class mapping to handle the case.

You're right and given that, I think we don't have to worry about https://github.com/kubernetes/kubernetes/pull/134326#discussion_r2455148605 as well, since the missing extended resources are filtered out in `removeMissingExtendedResources`.

sairameshv commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -422,7 +407,18 @@ func (f *Fit) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldO
 		return fwk.Queue, err
 	}
 	// Leaving in the queue, since the pod won't fit into the modified node anyway.
```
Addressed

sairameshv commented on pkg/scheduler/framework/plugins/noderesources/fit_test.go:
```diff
@@ -1797,7 +1795,7 @@ func TestHaveAnyRequestedResourcesIncreased(t *testing.T) {
 			}).Obj(),
 			draExtendedResourceEnabled: true,
-			expected:                   true,
+			expected:                   false,
```
This has been reverted with the introduction of draManager nil check while delegating the request to DRA

macsko commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -421,8 +416,18 @@ func (f *Fit) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldO
+	}
 	// Leaving in the queue, since the pod won't fit into the modified node anyway.
-	if !isFit(pod, modifiedNode, ResourceRequestsOptions{EnablePodLevelResources: f.enablePodLevelResources, EnableDRAExtendedResource: f.enableDRAExtendedResource}) {
+	if !isFit(pod, modifiedNode, opts, draManager) {
```
```suggestion
	if !isFit(pod, modifiedNode, draManager, opts) {
```

macsko commented on pkg/scheduler/framework/plugins/noderesources/fit_test.go:
```diff
@@ -632,17 +634,10 @@ func TestEnoughRequests(t *testing.T) {
-			nilResourceToDeviceClass:  true,
+			nodeInfo: framework.NewNodeInfo(newResourcePod(framework.Resource{MilliCPU: 0, Memory: 0})),
+			name:     "extended resource backed by DRA",
+			// When DRAExtendedResource is enabled and there's a matching DeviceClass,
```
I think the test with nil mapping (now - nil draManager) could stay here.

sairameshv replied:
Done

macsko commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,135 @@
+				"extendedResource", resourceName,
+				"deviceClass", deviceClass.Name)
+		}
+	}
```
Maybe just:
```suggestion
	delete(c.mapping, v1.ResourceName(*deviceClass.Spec.ExtendedResourceName))
	delete(c.mapping, v1.ResourceName(resourceapi.ResourceDeviceClassPrefix + deviceClass.Name))
```

macsko commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,135 @@
+}
+
+// updateMapping updates the cache with the device class mapping.
+func (c *ExtendedResourceCache) updateMapping(deviceClass *resourceapi.DeviceClass) {
```
What with the previous mapping for this device class? Is it mutable or not?

yliaog replied:
device class's extended resource name is mutable, we should delete the stale entry

sairameshv commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,135 @@
+		deviceClassLister := deviceClassInformer.Lister()
+		classes, err := deviceClassLister.List(labels.Everything())
+		if err != nil {
+			logger.Error(err, "Failed to populate extended resource cache during initialization")
```
I think this should be okay, similar to most of the constructors not returning error so as to not block the scheduler to start. If there is a failure listing the DeviceClasses here, the constructor would return an empty map initially and would be synced eventually when there are events getting registered(OnAdd, OnUpdate, OnDelete).
This would cause a temporary inconsistency during the startup but would eventually be synced.

Having said so, I'm also open to other thoughts on why we should strictly return an error.
cc: @yliaog

sairameshv commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache_test.go:
```diff
@@ -0,0 +1,99 @@
+	// Test modifying a device class
+	deviceClass1Modified := deviceClass1.DeepCopy()
+	deviceClass1Modified.Spec.ExtendedResourceName = ptr.To("test.com/gpu")
+	cache.OnUpdate(deviceClass1, deviceClass1Modified)
```
Makes sense

yliaog commented:
/retest

yliaog commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,135 @@
+
+// OnDelete handles deletion of a device class.
+func (c *ExtendedResourceCache) OnDelete(obj interface{}) {
+	if deviceClass, ok := obj.(*resourceapi.DeviceClass); ok {
```
informer may send DeletedFinalStateUnknown instead of the deviceClass, so this case also needs to be handled, please check out how other controllers handle it. It should look something like below:

	deviceClass, ok := obj.(*resourceapi.DeviceClass)
	if !ok {
		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
		if !ok {
			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
			return
		}
		deviceClass, ok = tombstone.Obj.(*resourceapi.DeviceClass)
		if !ok {
			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a DeviceClass %#v", obj))
			return
		}
	}

yliaog commented on pkg/kubelet/lifecycle/predicate.go:
```diff
@@ -388,7 +388,7 @@ func (e *PredicateFailureError) GetReason() string {
 // generalFilter checks a group of filterings that the kubelet cares about.
 func generalFilter(pod *v1.Pod, nodeInfo *schedulerframework.NodeInfo) []PredicateFailureReason {
-	admissionResults := scheduler.AdmissionCheck(pod, nodeInfo, true)
+	admissionResults := scheduler.AdmissionCheck(pod, nodeInfo, true, nil)
```
this change should not be needed, please revert.

yliaog commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -673,12 +681,13 @@ func preCheckForNode(nodeInfo *framework.NodeInfo) queue.PreEnqueueCheck {
 		EnablePodLevelResources:   utilfeature.DefaultFeatureGate.Enabled(features.PodLevelResources),
 		EnableDRAExtendedResource: utilfeature.DefaultFeatureGate.Enabled(features.DRAExtendedResource),
-	})
+	}, draManager)
```
you may pass nil in place of draManager here.

yliaog commented on pkg/scheduler/eventhandlers_test.go:
```diff
@@ -414,7 +415,11 @@ func TestPreCheckForNode(t *testing.T) {
+			sched := &Scheduler{
+				Profiles: profile.Map{},
+			}
+			preCheckFn := sched.preCheckForNode(nodeInfo)
```
i think the changes in this file could be reverted, right?

yliaog commented on test/integration/scheduler/queueing/queue.go:
```diff
@@ -433,6 +436,16 @@ var CoreResourceEnqueueTestCases = []*CoreResourceEnqueueTestCase{
+					ExtendedResourceName: ptr.To("example.com/gpu"),
+				},
+			},
+		},
```
why do you need this InitialDeviceClasses ? it was not needed before.

yliaog replied:
In testExtendedResource() before the test starts, you created a device class

	c, err := tCtx.Client().ResourceV1().DeviceClasses().Create(tCtx, testClass, metav1.CreateOptions{})

this device class should suffice? 

is it due to the race between the test, and informer update that causes: The global cache has no mapping for "example.com/gpu" ? what if you add a deley after the class creation, before the test start with something like:

time.Sleep(10*time.Seconds)

sairameshv replied:
The `testExtendedResource()` function and `TestCoreResourceEnqueue` are in completely different test files:

  1. `testExtendedResource()`: In `test/integration/dra/dra_test.go`
  2. `TestCoreResourceEnqueue`: In `test/integration/scheduler/queueing/queue.go`

I think, each test suite creates its own isolated environment:
  So the `DeviceClass` created in `testExtendedResource()` is NOT visible to `TestCoreResourceEnqueue`

yliaog replied:
I see. that makes sense.

yliaog commented on test/integration/dra/dra_test.go:
```diff
@@ -700,36 +700,56 @@ func testPrioritizedList(tCtx ktesting.TContext, enabled bool) {
-					"Type":   gomega.Equal(v1.PodScheduled),
-					"Status": gomega.Equal(v1.ConditionTrue),
-				}),
-			))
```
it was able to be scheduled before when enabled is true, why it cannot be scheduled anymore?

yliaog commented:
/retest

yliaog commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	// Populate the cache with existing device classes from the informer
+	if deviceClassInformer != nil {
```
it's easier to read by reduce the nesting

if deviceClassInformer == nil {
  return cache
}

yliaog commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,161 @@
+		deviceClassLister := deviceClassInformer.Lister()
+		classes, err := deviceClassLister.List(labels.Everything())
+		if err != nil {
+			logger.Error(err, "Failed to populate extended resource cache during initialization")
```
similarly, return early "return cache", then the following else can be removed

yliaog commented:
/lgtm

i left some nit comments, overall looks good 

@macsko could you please take another look?

sairameshv commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache_test.go:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	// Should not have the mapping for test.com/gpu
+	deviceClassName = cache.GetDeviceClass("test.com/gpu")
```
You are right Yu, Updated

macsko commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache_test.go:
```diff
@@ -0,0 +1,102 @@
+	}
+	if cache.GetDeviceClass("example.com/gpu") != "" {
+		t.Errorf("Expected 'example.com/gpu' to be removed after modification, got %s", cache.GetDeviceClass("example.com/gpu"))
+	}
```
Move the comment:
```suggestion
	// Should have the new mapping
	if cache.GetDeviceClass("test.com/gpu") != "gpu-class" {
		t.Errorf("Expected to find device class 'gpu-class' for 'test.com/gpu' after modification, got %s", deviceClassName)
	}
	// Should not have the old mapping for example.com/gpu
	if cache.GetDeviceClass("example.com/gpu") != "" {
		t.Errorf("Expected 'example.com/gpu' to be removed after modification, got %s", cache.GetDeviceClass("example.com/gpu"))
	}
```

macsko commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache_test.go:
```diff
@@ -0,0 +1,102 @@
+	deviceClassName = cache.GetDeviceClass("example.com/gpu")
+	if deviceClassName != "" {
+		t.Errorf("Expected 'example.com/gpu' to be removed after deleting device class, got %s", deviceClassName)
+	}
```
This check is unnecessary. Deletion of the old mapping was tested after update

macsko reviewed (commented):
Change looks correct. Small comments left

sairameshv commented:
Hello @yliaog @macsko @bart0sh 
I've rebased, resolved the merge conflicts with the recently merged [PR](https://github.com/kubernetes/kubernetes/pull/133784) and also addressed all the review comments.
Could you all PTAL at the changes ?

Thanks

yliaog commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,175 @@
+
+	if deviceClass != nil && deviceClass.Spec.ExtendedResourceName != nil {
+		delete(c.mapping, v1.ResourceName(*deviceClass.Spec.ExtendedResourceName))
+		delete(c.mapping, v1.ResourceName(resourceapi.ResourceDeviceClassPrefix+deviceClass.Name))
```
the implicit should be deleted outside deviceClass.Spec.ExtendedResourceName != nil, as long as deviceClass != nil

yliaog commented on pkg/scheduler/framework/plugins/noderesources/fit_test.go:
```diff
@@ -632,17 +635,21 @@ func TestEnoughRequests(t *testing.T) {
 			draExtendedResourceEnabled: true,
+			nilDRAManager:              true,
 			pod: newResourcePod(
 				framework.Resource{MilliCPU: 1, Memory: 1, ScalarResources: map[v1.ResourceName]int64{extendedResourceDRA: 1}}),
```
here you don't want to use extendedResourceDRA, instead, use extendedResourceA. otherwise, the global cache will have the mapping, and it will be satisfied.

sairameshv committed:
Add a global cache to support DRA's extended resource to the device
class mapping

- Add a new interface "DeviceClassResolver" in the scheduler framework
- Add a global cache of mapping between the extended resource and the
  device class
- Cache can be leveraged by the k8s api-server, controller-manager along with the scheduler
- This change helps in delegating the requests to the dynamicresource
  plugin based on the mapping during the node update events and thus
avoiding an extra scheduling cycle

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

yliaog commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,175 @@
+	// Remove old mappings first to handle ExtendedResourceName changes
+	if oldDeviceClass != nil && oldDeviceClass.Spec.ExtendedResourceName != nil {
+		delete(c.mapping, v1.ResourceName(*oldDeviceClass.Spec.ExtendedResourceName))
+		delete(c.mapping, v1.ResourceName(resourceapi.ResourceDeviceClassPrefix+oldDeviceClass.Name))
```
i see. in that case, it does not matter.

yliaog commented:
/lgtm

macsko reviewed (commented):
Change looks correct. Small comments left

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+	}
+	oldDeviceClass, ok := oldObj.(*resourceapi.DeviceClass)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("object is not a DeviceClass %#v", oldObj))
```
HandleErrorWithLogger

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
```
HandleErrorWithLogger

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+		}
+		deviceClass, ok = tombstone.Obj.(*resourceapi.DeviceClass)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a DeviceClass %#v", obj))
```
HandleErrorWithLogger

pohly reviewed (commented):
Please remove "Fix #133670" from the commit message. That won't help someone using `git log` and it spams the issue.

Why is this labeled as bug fix? It's a feature.

/label feature
/remove-label bug

Not a full review, need to look more closely.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+)
+
+// ExtendedResourceCache maintains a global cache of extended resource to device class mappings
+// that can also be accessed by event handlers without requiring an extra scheduling cycle.
```
"can be accessed by event handlers without requiring an extra scheduling cycle" sounds rather odd. Why should access need an extra cycle?

Here's an attempt to describe this thing:

```suggestion
// ExtendedResourceCache caches the mapping from extended resource name to
// device class. The mapping is updated based on informer events.
```

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+// NewExtendedResourceCache creates a new ExtendedResourceCache instance, populates it
+// with the current device classes from the informer cache, and registers itself as an
```
In practice this gets called in `NewDRAManager` at a time when the informers haven't been started yet. Therefore populating from the informer cache is a NOP and should be removed.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+	for _, deviceClass := range classes {
+		cache.updateMapping(deviceClass, nil)
+	}
+	logger.V(4).Info("Created and populated extended resource cache", "mappingCount", len(cache.mapping))
```
kube-scheduler.log:

```
I1029 10:20:44.289184  331892 cache.go:63] "Created and populated extended resource cache" mappingCount=0
```

Let's remove this initial listing and rely only on informer events.

pohly replied:
Solved in https://github.com/sairameshv/kubernetes/pull/1

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+	logger.V(4).Info("Created and populated extended resource cache", "mappingCount", len(cache.mapping))
+
+	// Register the cache as an event handler for device class changes
+	if _, err := deviceClassInformer.Informer().AddEventHandler(cache); err != nil {
```
There's no clean way to destroy an `ExtendedResourceCache` instance. You need to capture the handle here and add a `Stop` method which unregisters the handler, otherwise there's a goroutine leak.

That's also an opportunity to add a `ExtendedResourceCache.HasSynced` method. That method then could be passed up to the scheduler startup and ensure that scheduling is delayed until the cache is up-to-date.

That would be desirable, right?

yliaog replied:
the current code does full listing before registering the event handler, i suggest to do the full listing after registering the event handler. does that make the difference?

pohly commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -421,8 +415,18 @@ func (f *Fit) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldO
+	var draManager fwk.SharedDRAManager
+	if f.enableDRAExtendedResource {
+		draManager = f.handle.SharedDRAManager()
+	}
```
Why the feature gate check? Shouldn't `f.handle.SharedDRAManager` return nil when the feature gate is off?

sairameshv replied:
Feature-gate check can be removed here

sairameshv commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+	if deviceClassInformer == nil {
+		logger.V(4).Info("Created extended resource cache without initial population (deviceClassInformer is nil)")
+		return cache
+	}
```
IMHO, I don't really have a supporting argument for this. But, usually, informer may not be nil and this is just a defensive code for an imaginary(nil informer) case. I'm open to delete the code if it doesn't make sense

pohly replied:
This was only required for unit testing. Resolved in https://github.com/sairameshv/kubernetes/pull/1

sairameshv commented on pkg/scheduler/framework/plugins/noderesources/fit.go:
```diff
@@ -476,32 +480,23 @@ func haveAnyRequestedResourcesIncreased(pod *v1.Pod, originalNode, modifiedNode
-				return true
-			}
+		if shouldDelegateResourceToDRA(rName, modifiedNodeInfo, draManager, opts) {
+			return true
```
> Is my understanding correct that this is achieved by not delegating to DRA for extended resource names that aren't backed by DRA? The original code here didn't check that. shouldDelegateResourceToDRA now does. Correct?

Yes, you are right. I'll update the PR description. 
`shouldDelegateResourceToDRA` now delegates the resources to DRA plugin that are not backed by the Device Plugin which is not the case with the previous code

macsko commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
+1 to this. We need to be careful with races between event handlers and queueing events.

> Also, do DeviceClass changes trigger rescheduling as I said above, with or without this PR?

And that's a good point. `DeviceClass` changes should be registered in the `NodeResourcesFit` plugin.

yliaog replied:
https://github.com/kubernetes/kubernetes/pull/134991 opened to add deviceclass events to noderesources plugin

pohly replied:
The race can be solved by chaining event handlers instead of registering them in parallel.

For the sake of time I implemented that approach and also addressed various nits in https://github.com/sairameshv/kubernetes/pull/1. From the commit message:

    DRA extended resource: chain event handlers
    
    The cache and scheduler event handlers cannot be registered separately in the
    informer, that leads to a race (scheduler might schedule based on event before
    cache is updated). Chaining event handlers (cache first, then scheduler) avoids
    this.
    
    This also ensures that the cache is up-to-date before the scheduler
    starts (HasSynced of the handler registration for the cache is checked).
    
    Other changes:
    - renamed package to avoid clash with other "cache" packages
    - clarified nil handling
    - feature gate check before instantiating the cache
    - per-test logging
    - utilruntime.HandleErrorWithLogger
    - simpler cache.DeletedFinalStateUnknown

@sairameshv: if there's agreement on this revised approach and those other changes, then you can squash my commit into your branch.

pohly committed:
DRA extended resource: chain event handlers

The cache and scheduler event handlers cannot be registered separately in the
informer, that leads to a race (scheduler might schedule based on event before
cache is updated). Chaining event handlers (cache first, then scheduler) avoids
this.

This also ensures that the cache is up-to-date before the scheduler
starts (HasSynced of the handler registration for the cache is checked).

Other changes:
- renamed package to avoid clash with other "cache" packages
- clarified nil handling
- feature gate check before instantiating the cache
- per-test logging
- utilruntime.HandleErrorWithLogger
- simpler cache.DeletedFinalStateUnknown

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:
```diff
@@ -61,6 +63,10 @@ func NewDRAManager(ctx context.Context, claimsCache *assumecache.AssumeCache, re
 	}
 
+	deviceClassInformer := informerFactory.Resource().V1().DeviceClasses()
+	extCache := deviceclasscache.NewExtendedResourceCache(deviceClassInformer, logger)
```
Changed in https://github.com/sairameshv/kubernetes/pull/1

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go:
```diff
@@ -0,0 +1,178 @@
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a DeviceClass %#v", obj))
+			return
+		}
+	}
```
https://github.com/sairameshv/kubernetes/pull/1

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/deviceclass/extendedresourcecache/extendedresourcecache_test.go:
```diff
@@ -0,0 +1,212 @@
+	resourceapi "k8s.io/api/resource/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/test/utils/ktesting"
```
```suggestion
	"k8s.io/klog/v2/ktesting"
```

Wrong auto-import, the in-tree variant is forbidden here.

pohly commented:
@sairameshv: a few more test failures to sort out.

Overall I think this is now going in the right direction. I believe @macsko is keeping track of progress and will be ready to do the final review, but I don't think we have heard from SIG Autoscaling.

/assign @towca @jackfrancis

yliaog committed:
fixed unit test and integration test failures

Fix minor nits

Signed-off-by: Sai Ramesh Vanka <svanka@redhat.com>

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -756,9 +749,8 @@ func (pl *DynamicResources) filterExtendedResources(state *stateData, pod *v1.Po
-
 		allocatable, okScalar := nodeInfo.GetAllocatable().GetScalarResources()[rName]
-		_, okDynamic := state.draExtendedResource.resourceToDeviceClass[rName]
+		okDynamic := cache.GetDeviceClass(rName) != ""
```
Nit: `okDynamic` is not very descriptive. Perhaps `isBackedByDRA`?

Can also be done in a future refactoring PR (cc @alaypatel07).

pohly reviewed (commented):
I'm okay with merging this, but would like to leave the LGTM to someone else:
- I wrote part of the code.
- I don't have time for an in-depth line-by-line review (only called out a few nits while quickly scrolling through it).

yliaog commented:
/lgtm

@macsko @pohly @jackfrancis @towca could you please approve if this PR looks good to you?

we have a few other PRs depending on this PR,
we can follow up with a smaller PR to make futher changes.

pohly commented:
/approve

yliaog commented:
@jackfrancis could you please approve the PR if it looks good to you?

jackfrancis reviewed (approved):
/approve

towca commented on pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:
```diff
@@ -80,6 +86,16 @@ func (s *DefaultDRAManager) DeviceClasses() fwk.DeviceClassLister {
+//
+// That's okay, extendedresourcecache.ExtendedResourceCache.GetDeviceClass
+// returns the empty string if called for nil.
+func (s *DefaultDRAManager) DeviceClassResolver() fwk.DeviceClassResolver {
```
@jackfrancis Yeah, we'll need to update [the DRA Snapshot](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/simulator/dynamicresources/snapshot/snapshot.go) to implement `DeviceClassResolver()` based on the tracked DeviceClasses. If the implementation was more complex I'd advocate for extracting the common logic from `ExtendedResourceCache` so that Cluster Autoscaler can reuse it. But in this case it's just a very simple mapping so I think we can just reimplement it.

I filed https://github.com/kubernetes/autoscaler/issues/8742 to track this on Cluster Autoscaler side.
